### PR TITLE
Use macro matrix! in tests. Solve issue #54.

### PR DESCRIPTION
**Pull request by gcollura**
I hope it looks good.
I wanted to read the implementation and in the meantime I solved issue #54 .
If anything is missing, please tell me, I'll fix it.

### CODE REVIEW

File: tests/mat/mod.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@ fn test_l_triangular_solve_errs() {
     let a = Matrix::<f64>::new(0, 0, vec![]);
```
Comment (AtheMathmo) [inline tests/mat/mod.rs]: I think we can change this one too?

File: tests/mat/mod.rs [hunk 2]
```diff
@@ -16,13 +16,15 @@ fn test_u_triangular_solve_errs() {
...
```
Comment (AtheMathmo) [inline tests/mat/mod.rs]: And here (+ probably a few other empty matrix cases?)

**Comment by gcollura** [comment]
There's a bug with empty matrices, it cannot infer the type of the matrix (i32, f64, etc). This is the error:
```
$ cargo test
   Compiling rulinalg v0.3.4 (file:///xxxx/rulinalg)
error[E0282]: unable to infer enough type information about `_`
 --> <rulinalg macros>:2:37
  |
2 | { use $ crate :: matrix :: Matrix ; Matrix :: new ( 0 , 0 , vec ! [  ] ) } } ;
  |                                     ^^^^^^^^^^^^^ cannot infer type for `_`
tests/mat/mod.rs:17:13: 17:22 note: in this expansion of matrix! (defined in <rulinalg macros>)
  |
  = note: type annotations or generic parameter binding required

error: aborting due to previous error
```
This is why I left them as they were. Ideas?

**Comment by Andlon** [comment]
@gcollura,

If you write

```rust
let mat = matrix![];
```

then - depending on what follows the above code - Rust may not be able to know which type the Matrix has (after all, you have not told it!), so you may need to be explicit:

```rust
let mat: Matrix<f64> = matrix![];
```

Note that this may not always be necessary, in which case we would prefer to avoid the explicit type specification.

**Comment by gcollura** [comment]
Sorry for the missing ones.
I would leave some of the docs examples as they are and I would also leave such type of matrix declaration: `let a = Matrix::new(2, 3, vec![1.0; 6]); `, which are more compact, but still easily readable.
What do you think?

**Comment by AtheMathmo** [comment]
I think that makes sense! Thanks

**Comment by Andlon** [comment]
@AtheMathmo:

this kind of matrix construction seems to come up a lot, at least in our tests. Perhaps we should introduce a factory method like `Matrix::constant(rows, cols, value)`...?

**Comment by AtheMathmo** [comment]
@Andlon thay may be worthwhile, though the alternative is:

```rust
const * Matrix::ones(rows, cols)
```

+/- some explicit typing

**Comment by Andlon** [comment]
@AtheMathmo,
ah, I forgot about that. That's probably just as good. It's certainly better than the `let a = Matrix::new(rows, cols, vec![value; rows * cols]);` approach, since it's a lot more obvious!

**Comment by gcollura** [comment]
Is it by design the scalar multiplication only compiles when the scalar comes after the matrix such as: `let mat: Matrix<f64> = Matrix::ones(rows, cols) * 3.` and not `let mat: Matrix<f64> = 3. * Matrix::ones(rows, cols)`? I may address this issue in another PR.

**Comment by Andlon** [comment]
@gcollura,

Good catch, this seems to be an omission. At least, I couldn't find any implementation of `Mul<Matrix<T>> for f64` when searching the sources (I might have missed something though).

@AtheMathmo: Perhaps create a new issue so make sure we have implementations of all binary operations for both directions?

**Comment by AtheMathmo** [comment]
I believe there was a reason I didn't do this, maybe generics or some other restriction made it impossible? I'd be very happy if someone else could figure it out though.

**Comment by Andlon** [comment]
@AtheMathmo,

interesting. That might very well be! I keep getting tripped up by Rust's (implicit?) cross-crate rules. I'd like to look into this, but I think I better try to finish #71 first. Between thesis work and travelling I've been pretty busy lately. If anyone else wants to look into it, that would also be great!

**Comment by gcollura** [comment]
Now it should be good to go. If I still missed something, please tell me!
Thanks

**Comment by AtheMathmo** [comment]
It looks good at a glance @gcollura - thanks! I'll try to give it a thorough read later today.

@Andlon yes those get me too. I'm not sure if they interact here or not, but they might cause problems as `Mul` is defined out of crate and can be implemented on 'f64' (for example) in other crates. Of course 'Matrix' is internal, but I'm not sure how rustc handles that.

And no need to worry about any delays. In case you haven't noticed I'm also lagging behind, organizing visas and many other things for the start of my own PhD :).

**Comment by Andlon** [comment]
@gcollura, looks good to me! Appreciate that you took the effort to do this!

@AtheMathmo, sounds exciting! Hope all goes well with the preparations and whatnot :)

File: src/matrix/decomposition/eigen.rs [hunk 3]
```diff
@@ -285,7 +285,8 @@ impl<T: Any + Float + Signed> Matrix<T> {
             }
 
             let (c, s) = Matrix::givens_rot(x, y);
-            let givens_mat = Matrix::new(2, 2, vec![c, -s, s, c]);
+            let givens_mat = matrix!(c, -s;
```
Comment (AtheMathmo) [inline src/matrix/decomposition/eigen.rs]: This is the only, incredibly minor, thing that I noticed. I don't mind this change in itself, but we keep the `Matrix::new` construction on the givens matrices in the `svd` module. I think I prefer slightly here to keep it as the `Matrix::new`. Either way my opinion here isn't particularly strong and I'll happily merge as is.

**Pull request merged**